batching_kafka_consumer/__init__.py

```diff
@@ -101,14 +101,16 @@ def __init__(self, topics, worker, max_batch_size, max_batch_time, bootstrap_ser
         self.worker = worker
 
         self.max_batch_size = max_batch_size
-        self.max_batch_time = max_batch_time
+        self.max_batch_time = max_batch_time  # in milliseconds
         self.metrics = metrics
         self.group_id = group_id
 
         self.shutdown = False
-        self.batch = []
-        self.uncommitted_messages = False
-        self.timer = None
+
+        self.__batch_results  = []
+        self.__batch_deadline = None
+        self.__batch_messages_processed_count = 0
+        self.__batch_messages_processed_duration_ms = 0.0  # in milliseconds
```

Comment:
I'd like to encapsulate these differently into a `Batch` data structure that has an `add` and `flush` method. This will make it easier to implement streaming writes. Right now this is too coupled to the idea that a batch is buffered in memory before writing.

- For the current implementation, `add` will append to the in-memory buffer, and `flush` will write the contents of the in-memory buffer to the socket.
- For the streaming implementation, `add` will write to the socket buffer instead, and `flush` will close the connection (or request, if we can keep-alive.)

Reply:
I'd like we found a better name that showed that this is the sum of all the processing times.
What about __batch_processing_time  ?

Reply:
So, long answer: in the past, we've talked about parallelizing the message processing to that we can increase  consumer throughput. These consumers are largely CPU-bound, and a large amount of that CPU is being spent deserializing messages for processing and serializing writes. Right now, since all partitions assigned to a consumer are processed in the same process, we can add consumers until there are the same numbers of consumer processes as partitions. This leads to uneven resource utilization (since messages are not uniformly distributed across partitions) and still limits our theoretical per-partition throughput to whatever can be handled by a single CPU, which can be a problem when we have large spikes or an accumulated backlog in a single partition.

So, to parallelize it, we're going to need to:

1. send the raw message (still serialized) to the subprocess
2. then, in the subprocess:
    1. receive deserialize the message,
    2. do all of the message processing,
    3. serialize the processed result,
3. receive the (re-)serialized message and prepare it for write.

If we're able to perform streaming writes (e.g. using the [JsonEachRow input format](https://clickhouse.yandex/docs/en/interfaces/formats/#jsoneachrow) with `HTTP Transfer-Encoding: chunked`), we can take the result yielded by the subprocess and just place the bytes directly onto the socket without needing to do any additional processing. If we stuck with the existing model, we'd need to have an additional step where we prepare the write by deserializing it back into Python structures just to serialize it back into the format used by the driver to write the batch.

Of course, there are also memory benefits to this as well (since we wouldn't have to accumulate an entire batch in memory in the parent/writer process) but this hasn't really been an issue for us up to this point.

We've also had issues with bugs in the driver before (in the native format implementation), and we suspect that having the JSON serialization occurring in a C extension is going to be more efficient than the Python implementation of the native extension.

Reply:
ahhhh you were not talking about Kafka streams.

Reply:
> ahhhh you were not talking about Kafka streams.

Oops, no — I forgot "streaming" was a little overloaded here.

> Got it. Does clickhouse still writes in batches with the JsonEachRow input format or would it be like doing running one separate INSERT per message?.

I'm 99% sure it does batching as if the entire payload was passed at once (I am pretty sure we tested this before) but it's probably worth double-checking again just to be absolutely sure.

> Also, please see my second comment about the naming.

Yep, this was handled in 0f285aed374e216f52e3a49e0aefb4b23b85f7aa (I didn't want to mark the conversation resolved to make sure the rest didn't get missed.)

Reply:
ok